Build soci-snapshotter package

larvacea opened the pull request:
**Description of changes:**
This builds the soci-snapshotter containerd plugin. The plugin has two components:

- `soci-snapshotter-grpc` is a daemon (systemd service).
- `soci` is a CLI tool for working with SOCI indices. 

In addition, the rpm packages (and should install) `soci-snapshotter.service`, the unit file for systemd, so the daemon starts during boot (after network, and before containerd).

The aws-dev variant (and no other variant) includes the package.

**Testing done:**
[Edit] Manual testing on an aws-dev ec2 instance, following the soci-snapshotter installation and getting-started guides. The soci-snapshotter documentation uses nerdctl, so these packages were tested using nerdctl. That required a few extra steps:

* Create an overlay file system on the /root directory because nerdctl wants to store registry credentials in $HOME/.docker/config.json.
* Install wget and tar in the admin container, download, then copy nerdctl over to the host.
* Add the soci-snapshotter plugin to containerd's configuration (in the AMI)

Following the steps outlined in the soci-snapshotter repo, I can verify containerd will happily pull and run an image through soci. I see exactly the transactions I expect with ECR and the plugin is able to find and download the soci index, ztoc, and manifest files given the package URI. Both pull and run benefit from lazy loading on the test image they suggested.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

yeazelm commented on packages/soci-snapshotter/Cargo.toml:
```diff
@@ -0,0 +1,30 @@
+[package]
+name = "soci-snapshotter"
+version = "0.5.0"
```
nit: we typically use 0.1.0 for the Cargo.toml version.

yeazelm commented on packages/soci-snapshotter/Cargo.toml:
```diff
@@ -0,0 +1,30 @@
+[package]
+name = "soci-snapshotter"
+version = "0.5.0"
+publish = false
```
We should include the `edition`

yeazelm commented on packages/soci-snapshotter/Cargo.toml:
```diff
@@ -0,0 +1,30 @@
+releases-url = "https://github.com/awslabs/soci-snapshotter/releases"
+
+[[package.metadata.build-package.external-files]]
+url = "https://github.com/larvacea/soci-snapshotter/archive/refs/tags/v0.5.0.tar.gz"
```
Use the upstream url for these: https://github.com/awslabs/soci-snapshotter

yeazelm commented on packages/soci-snapshotter/soci-snapshotter.spec:
```diff
@@ -0,0 +1,54 @@
+%global rpmver %{gover}
+
+Name: %{_cross_os}soci-snapshotter
+Version: 0.5.0
```
Can we use the `gover` from above so updating is a single place?

yeazelm commented on packages/soci-snapshotter/soci-snapshotter.spec:
```diff
@@ -0,0 +1,54 @@
+Name: %{_cross_os}soci-snapshotter
+Version: 0.5.0
+Release: 1%{?dist}
+Summary: Snapshots things
```
Is there a better summary than this?

yeazelm commented on packages/soci-snapshotter/soci-snapshotter.spec:
```diff
@@ -0,0 +1,54 @@
+Summary: Snapshots things
+License: Apache-2.0
+URL: https://github.com/awslabs/soci-snapshotter
+Source0: https://github.com/larvacea/soci-snapshotter/archive/refs/tags/v0.5.0.tar.gz
```
Can we use `%{gover}` or similar to make this easier to update in the future?

yeazelm commented on packages/soci-snapshotter/soci-snapshotter.spec:
```diff
@@ -0,0 +1,54 @@
+License: Apache-2.0
+URL: https://github.com/awslabs/soci-snapshotter
+Source0: https://github.com/larvacea/soci-snapshotter/archive/refs/tags/v0.5.0.tar.gz
+Source1: bundled-v0.5.0.tar.gz
```
Can we use `%{gover}` or similar to make this easier to update in the future?

yeazelm commented on packages/soci-snapshotter/soci-snapshotter.spec:
```diff
@@ -0,0 +1,54 @@
+Source101: soci-snapshotter.service
+Source1000: clarify.toml
+
+BuildRequires: %{_cross_os}glibc-devel %{_cross_os}libz-devel
```
nit: Make these distinct lines for `BuildRequires:`

yeazelm reviewed (commented):
Per your testing comments:
> Absolutely none. This definitely builds the RPM, and the rpm tools say it is all perfectly fine. The aws-dev variant builds. If one wanted to inform containerd of the plugin's existence, one would need to add a few lines to the containerd config.toml file.

Can we try a bit more validation than this? Since it is a dev image, I don't think we need to fully integrate the API for the configuration settings, but at least you could attempt to get containerd to use it via the shell manually? Just because it builds into an rpm doesn't mean the resulting binaries are usable. I'd like to at least see a basic smoke test to ensure the binaries could work once the rest of the configuration is provided.

arnaldo2792 commented on packages/soci-snapshotter/soci-snapshotter.spec:
```diff
@@ -0,0 +1,54 @@
+
+%set_cross_go_flags
+
+(cd cmd ; go build -buildmode=pie -ldflags="${GOLDFLAGS}" -o "../out/soci-snapshotter-grpc" ./soci-snapshotter-grpc)
+(cd cmd ; go build -buildmode=pie -ldflags="${GOLDFLAGS}" -o "../out/soci" ./soci)
```
We don't start subshells when we build multiple binaries. Instead of `cd`, you can tell `go` to compile from a path. See [this] for reference.

[this]: https://github.com/bottlerocket-os/bottlerocket/blob/507c21fa421f073861480be967ea8d053561e18b/packages/amazon-ssm-agent/amazon-ssm-agent.spec#L30

arnaldo2792 commented on packages/soci-snapshotter/soci-snapshotter.spec:
```diff
@@ -0,0 +1,54 @@
+%cross_scan_attribution --clarify %{S:1000} go-vendor vendor
+
+%files
+%license LICENSE
+%license NOTICE.md
```
```suggestion
%license LICENSE NOTICE.md
```

arnaldo2792 commented on packages/soci-snapshotter/soci-snapshotter.service:
```diff
@@ -0,0 +1,42 @@
+Restart=always
+RestartSec=5
+
+[Install]
+WantedBy=multi-user.target
```
I think we do want the new line at the end of the file?

bcressey replied:
Does containerd have a hard dependency on this service if configured to use the SOCI snapshotter?
```suggestion
[Install]
WantedBy=multi-user.target
RequiredBy=containerd.service
```

arnaldo2792 commented on packages/soci-snapshotter/soci-snapshotter.service:
```diff
@@ -0,0 +1,42 @@
+# Copyright The Soci Snapshotter Authors.
```
Isn't this service in the pulled sources? If it is, you can just copy it from them, instead of creating a new service.

larvacea replied:
The upstream has this file but the paths are different:
```
37c37
< ExecStart=/usr/bin/soci-snapshotter-grpc
---
> ExecStart=/usr/local/bin/soci-snapshotter-grpc

```
So close, but no.

arnaldo2792 commented on packages/soci-snapshotter/Cargo.toml:
```diff
@@ -0,0 +1,30 @@
+[build-dependencies]
+glibc = { path = "../glibc" }
+libz = { path = "../libz" }
+[dependencies]
```
New line here to be consistent with other packages, see [this]

[this]: https://github.com/bottlerocket-os/bottlerocket/blob/507c21fa421f073861480be967ea8d053561e18b/packages/containerd/Cargo.toml#L22

arnaldo2792 commented on packages/fuse/fuse.spec:
```diff
@@ -0,0 +1,93 @@
+%description
+With FUSE it is possible to implement a fully functional filesystem in a
+userspace program. This package contains the FUSE v2 userspace tools to
+mount a FUSE filesystem.
```
By convention, we just use the summary in all the descriptions.

arnaldo2792 commented on packages/fuse/fuse.spec:
```diff
@@ -0,0 +1,93 @@
+# https://bugzilla.redhat.com/show_bug.cgi?id=1984776
+Patch7: fuse2-0007-util-ulockmgr_server.c-conditionally-define-closefro.patch
+
+Conflicts:	filesystem < 3
```
We don't care about the version of the filesystem package, so you can avoid this.

arnaldo2792 commented on packages/fuse/fuse.spec:
```diff
@@ -0,0 +1,93 @@
+Summary:	File System in Userspace (FUSE) v2 utilities
+License:	GPL-1.0-or-later
+URL:		http://fuse.sf.net
+Source0:	https://github.com/libfuse/libfuse/releases/download/fuse-2.9.9/fuse-2.9.9.tar.gz
```
You can use `%{version}` here instead of hard-coding it.

arnaldo2792 commented on packages/fuse/fuse.spec:
```diff
@@ -0,0 +1,93 @@
+Name:		%{_cross_os}fuse
```
It is ok to not align all these with tabs, you can use a single space.

arnaldo2792 commented on packages/fuse/fuse.spec:
```diff
@@ -0,0 +1,93 @@
+Patch7: fuse2-0007-util-ulockmgr_server.c-conditionally-define-closefro.patch
+
+Conflicts:	filesystem < 3
+BuildRequires:	libselinux-devel
```
If this requires selinux, we need to use `%{_cross_os}libselinux-devel`

arnaldo2792 commented on packages/fuse/fuse.spec:
```diff
@@ -0,0 +1,93 @@
+
+Conflicts:	filesystem < 3
+BuildRequires:	libselinux-devel
+BuildRequires:	autoconf, automake, libtool, gettext-devel, make
```
I think you can skip these, do you get an error if you don't?

arnaldo2792 commented on packages/fuse/fuse.spec:
```diff
@@ -0,0 +1,93 @@
+userspace program. This package contains the FUSE v2 userspace tools to
+mount a FUSE filesystem.
+
+%package libs
```
We use `devel` for development libraries

arnaldo2792 commented on packages/fuse/fuse.spec:
```diff
@@ -0,0 +1,93 @@
+userspace program. This package contains the FUSE v2 libraries.
+
+%prep
+%autosetup -n fuse-2.9.9 -p 1
```
```suggestion
%autosetup -n %{name}-%{version} -p 1
```

arnaldo2792 commented on packages/fuse/Cargo.toml:
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "fuse"
+version = "2.9.9"
```
According to the GitHub repo, this is a maintance version, and folks should move to the latest 3.x version. I didn't see any requirements on the version for fuse in SOCI, but it will be best to ask the developers to make sure we meet SOCI's requirements.

arnaldo2792 commented on packages/fuse/fuse.spec:
```diff
@@ -0,0 +1,93 @@
+%build
+# Can't pass --disable-static here, or else the utils don't build
+export MOUNT_FUSE_PATH="%{_sbindir}"
+CFLAGS="%{optflags} -D_GNU_SOURCE" %configure
```
We have our own CFLAGS, you can set them with `%set_cross_build_flags`

arnaldo2792 commented on packages/fuse/fuse.spec:
```diff
@@ -0,0 +1,93 @@
+
+%build
+# Can't pass --disable-static here, or else the utils don't build
+export MOUNT_FUSE_PATH="%{_sbindir}"
```
If this has to be installed at `sbindir`, use `%{_cross_sbindir}`

arnaldo2792 commented on packages/fuse/fuse.spec:
```diff
@@ -0,0 +1,93 @@
+# Can't pass --disable-static here, or else the utils don't build
+export MOUNT_FUSE_PATH="%{_sbindir}"
+CFLAGS="%{optflags} -D_GNU_SOURCE" %configure
+make %{?_smp_mflags}
```
You could just use `%make_build`

arnaldo2792 commented on packages/fuse/fuse.spec:
```diff
@@ -0,0 +1,93 @@
+# Get rid of static libs
+rm -f %{buildroot}%{_cross_libdir}/*.a
+
+%ldconfig_scriptlets libs
+
+ls -l %{buildroot}%{_cross_bindir}
+ls -l %{buildroot}%{_cross_sbindir}
+ls -l %{buildroot}%{_cross_libdir}
+ls -l %{buildroot}%{_cross_licensedir}
```
You shouldn't need this.

larvacea replied:
Oh drat, my debugging cruft!

arnaldo2792 commented on packages/fuse/fuse.spec:
```diff
@@ -0,0 +1,93 @@
+%license COPYING
+%{_cross_attribution_file}
+%{_cross_sbindir}/mount.fuse
+%attr(4755,root,root) %{_cross_bindir}/fusermount
```
You should only have `%{_cross_bindir}/fusermount`.

larvacea replied:
Should I? The fusermount executable is installed set-uid-root. Anything that tries to fork a fusermount on bottlerocket is likely to already be root, so maybe we don't need to make the executable suid-root. What do you think?

arnaldo2792 replied:
I think you are right, because the systemd services will be running as UID 0.

arnaldo2792 commented on packages/fuse/fuse.spec:
```diff
@@ -0,0 +1,93 @@
+%prep
+%autosetup -n fuse-2.9.9 -p 1
+
+# ./makeconf.sh
```
If we do use v2.9.9, we should change how to compile the package. I've been playing with it and I think we should do something like this:

```
./makeconf.sh
./configure --prefix=%{_cross_prefix} #... and all other flags
```

And disable all the things we don't want like `--disable-utils` and `--disable-static`. Then, since you used the correct prefixes, you could just `%make_install` in the install section.

larvacea replied:
soci-snapshotter's main dependency is the fusermount executable from util, so we need --enable-utils. The only thing that soci-snapshotter does (directly) is query whether there is an executable fusermount on its path; it looks at the error (if any) from the query and throws away the information about the executable itself. I have not looked into all of the dependencies to see if they, then, do anything with fusermount.

larvacea commented:
Testing confirms that soci-snapshotter (running on bottlerocket, as a systemd service) works with or without fuse. Without fuse, soci-snapshotter will write error messages to the journal.

bcressey commented on packages/soci-snapshotter/soci-snapshotter.service:
```diff
@@ -0,0 +1,42 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
Over time we usually end up re-writing these unit files so heavily that I'm not sure the license makes sense. But in any case this header can be omitted since it's covered by the larger license at the project root.

bcressey commented on packages/soci-snapshotter/soci-snapshotter.service:
```diff
@@ -0,0 +1,42 @@
+[Unit]
+Description=soci snapshotter containerd plugin
+Documentation=https://github.com/awslabs/soci-snapshotter
+After=network.target
+Before=containerd.service
```
`network.target` is kind of a useless dependency since it doesn't imply that the network is up. Also, an ordering dependency in the absence of another dependency is generally incomplete.

For Bottlerocket we have `configured.target` to cover the majority of initial system configuration and to prevent units from being scheduled until first boot configuration is complete and bootstrap containers have a chance to run:
```suggestion
After=configured.target
Wants=configured.target
Before=containerd.service
```

bcressey commented on packages/soci-snapshotter/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+releases-url = "https://github.com/awslabs/soci-snapshotter/releases"
+
+[[package.metadata.build-package.external-files]]
+url = "https://github.com/awslabs/soci-snapshotter/archive/refs/tags/v0.5.0.tar.gz"
```
I prefer using URLs of this form so that the downloaded file has the expected name, rather than just `v0.5.0.tar.gz`:
```suggestion
url = "https://github.com/awslabs/soci-snapshotter/archive/v0.5.0/soci-snapshotter-0.5.0.tar.gz"
```

bcressey commented on packages/soci-snapshotter/soci-snapshotter.spec:
```diff
@@ -0,0 +1,53 @@
+Summary: A containerd snapshotter plugin which enables lazy loading for OCI images.
+License: Apache-2.0
+URL: https://github.com/awslabs/soci-snapshotter
+Source0: https://github.com/awslabs/soci-snapshotter/archive/refs/tags/v0.5.0.tar.gz
```
Reuse the macros and use the other URL, so that only the macro at the top needs to be updated in future spec edits:
```suggestion
Source0: https://github.com/awslabs/%{gorepo}/archive/%{gover}/%{gorepo}-%{gover}.tar.gz
```

bcressey commented on packages/soci-snapshotter/soci-snapshotter.spec:
```diff
@@ -0,0 +1,53 @@
+License: Apache-2.0
+URL: https://github.com/awslabs/soci-snapshotter
+Source0: https://github.com/awslabs/soci-snapshotter/archive/refs/tags/v0.5.0.tar.gz
+Source1: bundled-v%{gover}.tar.gz
```
```suggestion
Source1: bundled-%{gorepo}-%{gover}.tar.gz
```

bcressey commented on packages/soci-snapshotter/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+bundle-modules = [ "go" ]
+
+[[package.metadata.build-package.external-files]]
+url = "https://github.com/awslabs/soci-snapshotter/archive/refs/tags/v0.5.0.tar.gz"
```
```suggestion
url = "https://github.com/awslabs/soci-snapshotter/archive/v0.5.0/soci-snapshotter-0.5.0.tar.gz"
```

bcressey commented on packages/soci-snapshotter/soci-snapshotter.spec:
```diff
@@ -0,0 +1,53 @@
+%global gorepo soci-snapshotter
```
To verify that this builds correctly with FIPS and non-FIPS binaries, after rebasing on develop:
```suggestion
%global cross_check_fips 1

%global gorepo soci-snapshotter
```

bcressey commented on packages/soci-snapshotter/soci-snapshotter.spec:
```diff
@@ -0,0 +1,53 @@
+
+%set_cross_go_flags
+
+go build -C cmd -buildmode=pie -ldflags="${GOLDFLAGS}" -o "../out/soci-snapshotter-grpc" ./soci-snapshotter-grpc
+go build -C cmd -buildmode=pie -ldflags="${GOLDFLAGS}" -o "../out/soci" ./soci
```
`-buildmode=pie` is redundant and should be omitted:
```suggestion
go build -C cmd -ldflags="${GOLDFLAGS}" -o "../out/soci-snapshotter-grpc" ./soci-snapshotter-grpc
gofips build -C cmd -ldflags="${GOLDFLAGS}" -o "../fips/soci-snapshotter-grpc" ./soci-snapshotter-grpc

go build -C cmd -ldflags="${GOLDFLAGS}" -o "../out/soci" ./soci
gofips build -C cmd -ldflags="${GOLDFLAGS}" -o "../fips/soci" ./soci
```

bcressey commented on packages/soci-snapshotter/soci-snapshotter.spec:
```diff
@@ -0,0 +1,53 @@
+%setup -T -D -n %{gorepo}-%{gover} -b 2 -q
+
+%build
+
```
nit: no need for empty line at start of section
```suggestion
```

bcressey commented on packages/soci-snapshotter/soci-snapshotter.spec:
```diff
@@ -0,0 +1,53 @@
+go build -C cmd -buildmode=pie -ldflags="${GOLDFLAGS}" -o "../out/soci" ./soci
+
+%install
+install -d %{buildroot}%{_cross_bindir}
```
```suggestion
install -d %{buildroot}{%{_cross_bindir},%{_cross_fips_bindir}}
```

bcressey commented on packages/soci-snapshotter/soci-snapshotter.spec:
```diff
@@ -0,0 +1,53 @@
+%install
+install -d %{buildroot}%{_cross_bindir}
+install -d %{buildroot}%{_cross_unitdir}
+install -p -m 0755 out/soci-snapshotter-grpc %{buildroot}%{_cross_bindir}
+install -p -m 0755 out/soci %{buildroot}%{_cross_bindir}
```
```suggestion
install -p -m 0755 out/soci-snapshotter-grpc %{buildroot}%{_cross_bindir}
install -p -m 0755 out/soci %{buildroot}%{_cross_bindir}
install -p -m 0755 fips/soci-snapshotter-grpc %{buildroot}%{_cross_fips_bindir}
install -p -m 0755 fips/soci %{buildroot}%{_cross_fips_bindir}
```

bcressey commented on packages/soci-snapshotter/soci-snapshotter.spec:
```diff
@@ -0,0 +1,53 @@
+%cross_scan_attribution --clarify %{S:1000} go-vendor vendor
+
+%files
+%license LICENSE
+%license NOTICE.md
```
nit: it's somewhat conventional to combine these:
```suggestion
%license LICENSE NOTICE.md
```

bcressey commented on packages/soci-snapshotter/soci-snapshotter.spec:
```diff
@@ -0,0 +1,53 @@
+Source1000: clarify.toml
+
+BuildRequires: %{_cross_os}glibc-devel
+BuildRequires: %{_cross_os}libz-devel
```
```suggestion
BuildRequires: %{_cross_os}libz-devel

Requires: %{name}(binaries)
```

bcressey commented on packages/soci-snapshotter/soci-snapshotter.spec:
```diff
@@ -0,0 +1,53 @@
+BuildRequires: %{_cross_os}glibc-devel
+BuildRequires: %{_cross_os}libz-devel
+
+%description
+%{summary}.
```
boilerplate for FIPS:
```suggestion
%description
%{summary}.

%package bin
Summary: SOCI snapshotter binaries
Provides: %{name}(binaries)
Requires: (%{_cross_os}image-feature(no-fips) and %{name})
Conflicts: (%{_cross_os}image-feature(fips) or %{name}-fips-bin)

%description bin
%{summary}.

%package fips-bin
Summary: SOCI snapshotter binaries, FIPS edition
Provides: %{name}(binaries)
Requires: (%{_cross_os}image-feature(fips) and %{name})
Conflicts: (%{_cross_os}image-feature(no-fips) or %{name}-bin)

%description fips-bin
%{summary}.
```

bcressey commented on packages/soci-snapshotter/soci-snapshotter.spec:
```diff
@@ -0,0 +1,53 @@
+%files
+%license LICENSE
+%license NOTICE.md
+%{_cross_bindir}/soci-snapshotter-grpc
+%{_cross_bindir}/soci
+%{_cross_unitdir}/soci-snapshotter.service
+%{_cross_attribution_vendor_dir}
+%{_cross_attribution_file}
```
```suggestion
%{_cross_unitdir}/soci-snapshotter.service
%{_cross_attribution_vendor_dir}
%{_cross_attribution_file}

%files bin
%{_cross_bindir}/soci-snapshotter-grpc
%{_cross_bindir}/soci

%files fips-bin
%{_cross_fips_bindir}/soci-snapshotter-grpc
%{_cross_fips_bindir}/soci
```

bcressey reviewed (commented):
Pretty sure the snapshotter binaries will need to be built for FIPS and non-FIPS, so I've suggested the changes required for that.

larvacea committed:
Add soci-snapshotter to bottlerocket and aws-dev

Add the SOCI snapshotter package to aws-dev. This is a lazy-loading plugin for
containerd.

See:

https://github.com/awslabs/soci-snapshotter

larvacea commented:
I have incorporated all of Ben's suggestions except changing the source URL. Changing the source URL broke the build, and I have not yet debugged that problem.

ducminhle commented:
Hi @larvacea,
Is this PR ready for EKS?
When I install soci-snapshotter on EKS (with pre-boostrap user-data), I need to install amazon-ecr-credential-helper and config $HOME/.docker/config.json to pull images from ECR.
```
{
  "credsStore": "ecr-login"
}
```

arnaldo2792 commented:
Hey @ducminhle , this is not ready to be used by EKS or ECS yet, due to the dependencies that you described. We need to work closely with the EKS and SOCI maintainers to figure out if there is a different solution to support credentials in SOCI.

ducminhle commented:
> Hey @ducminhle , this is not ready to be used by EKS or ECS yet, due to the dependencies that you described. We need to work closely with the EKS and SOCI maintainers to figure out if there is a different solution to support credentials in SOCI.

Thank you for your answer
